### PR TITLE
feat: trainees list filter

### PR DESCRIPTION
**Pull request by omarCreativeDev**
TISNEW-4154 - implemented trainees list filtering

- added new `AllDoctorsFilter` and `UnderNoticeFilter` actions, selectors and related events
- added new `TraineeFiltersComponent` alongside some minor ui/scss tweaks
- extracted generation of http params in to service so code can be re used and easier unit testing
- tweaked and added more unit tests

### CODE REVIEW

File: src/app/core/trainee/trainee.service.ts [hunk 1]
```diff
@@ -24,6 +24,22 @@ export class TraineeService {
...
   }
 
+  public generateParams(): HttpParams {
+    const snapshot: TraineesStateModel = this.store.snapshot().trainees;
+
+    let params: HttpParams = new HttpParams()
+      .set("sortColumn", snapshot.sort.active)
+      .set("sortOrder", snapshot.sort.direction)
+      .set("pageNumber", snapshot.pageIndex.toString())
+      .set("underNotice", snapshot.underNotice.toString());
```
Comment (ojemuyiwa) [inline src/app/core/trainee/trainee.service.ts:34]: wrapping this like search query seems like the right approach to adding under notice filter or not

File: src/app/trainees/trainee-filters/trainee-filters.component.html [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<ng-container
+  *ngIf="{
+    countTotal: countTotal$ | async,
+    countUnderNotice: countUnderNotice$ | async,
+    underNotice: underNotice$ | async
+  } as data"
+>
+  <button
```
Comment (ojemuyiwa) [inline src/app/trainees/trainee-filters/trainee-filters.component.html:8]: which button is active and which is in-active state wise? mat-tabs or routeActiveLink may be a better way to go for uniformity, but whichever way we go lets have an active and inactive state so users know what filter is active.

File: src/app/core/trainee/trainee.service.ts (hunk 1 above)
Comment (omarCreativeDev) [reply src/app/core/trainee/trainee.service.ts:34]: cheers for flagging, just double checked with BE on team reval slack and underNotice=false is the way to return all doctors

File: src/scss/utilities/_spacing-helpers.scss [hunk 3]
```diff
@@ -19,7 +19,7 @@ $space-prefixes: (
   @each $attr-short, $attr-long in $space-prefixes {
     @each $value in $space-values {
       .#{$attr-short}-#{$value} {
-        #{$attr-long}: #{$value}#{"px"};
+        #{$attr-long}: #{$value}#{"px"} !important;
```
Comment (ojemuyiwa) [inline src/scss/utilities/_spacing-helpers.scss:22]: important is not usually recommended maybe we can increase css priority if something needs overriding?

File: src/app/trainees/trainee-filters/trainee-filters.component.html (hunk 2 above)
Comment (omarCreativeDev) [reply src/app/trainees/trainee-filters/trainee-filters.component.html:8]: ah yea good point but am setting the active state via the `mat-primary` class which is bound to `underNotice` true|false from store

File: src/scss/utilities/_spacing-helpers.scss (hunk 3 above)
Comment (omarCreativeDev) [reply src/scss/utilities/_spacing-helpers.scss:22]: yea good point. the !important is very ugly but had to override some out of the box angular material css. will investigate a heavier css selector perhaps as that would be cleaner

File: src/app/core/trainee/trainee.service.ts (hunk 1 above)
Comment (ojemuyiwa) [reply src/app/core/trainee/trainee.service.ts:34]: my guess with working with filters underNotice=false will return a sub-set of doctors where underNotice=false i.e. Not All Doctors

Comment (omarCreativeDev) [reply src/app/core/trainee/trainee.service.ts:34]: as far as I know from uzair `underNotice=false` returns all doctors. is that correct @uzairbaig?

File: src/app/trainees/reset-trainee-list/reset-trainee-list.component.html [hunk 4]
```diff
@@ -1 +1,3 @@
-<span class="cursor-pointer" (click)="resetTraineeList()">Clear all</span>
+<button mat-button (click)="resetTraineeList()" aria-label="Clear all">
+  Clear all
```
Comment (ojemuyiwa) [inline src/app/trainees/reset-trainee-list/reset-trainee-list.component.html:2]: Seems the reset may need to be revised if underNotice=false isn't default i.e. only if

File: src/app/trainees/reset-trainee-list/reset-trainee-list.component.ts [hunk 5]
```diff
@@ -24,9 +24,9 @@ export class ResetTraineeListComponent {
   constructor(private store: Store, private traineeService: TraineeService) {}
```
Comment (ojemuyiwa) [inline src/app/trainees/reset-trainee-list/reset-trainee-list.component.ts:24]: The clear / reset functionality seems not to set to default which I thought was `sortColumn=submissionDate&sortOrder=desc&pageNumber=0`

Comment (ojemuyiwa) [inline src/app/trainees/reset-trainee-list/reset-trainee-list.component.ts:24]: The clear / reset functionality seems not to set to default which I thought was sortColumn=submissionDate&sortOrder=desc&pageNumber=0

File: src/scss/utilities/_spacing-helpers.scss (hunk 3 above)
Comment (ojemuyiwa) [reply src/scss/utilities/_spacing-helpers.scss:22]: let me know if you want to override in component itself or revise this later as in the grand scheme of things is not a show stopper.

Comment (omarCreativeDev) [reply src/scss/utilities/_spacing-helpers.scss]: !important removed and pr updated

**Pull request merged**